lib/providers/ipmeta_provider_maxmind_v2.c

```diff
@@ -0,0 +1,914 @@
+    tmp->id = strtol(tok, &end, 10);
+    if (end == tok || *end != '\0' || errno == ERANGE) {
+      ipmeta_log(__func__, "Invalid ID Value (%s)", tok);
+      state->parser.status = CSV_EUSER;
+    }
+    }
+    break;
+
+  case BLOCKS_COL_REPRESENTED_CCGEONAME_ID:
+    /* skip column */
+    break;
+
+  case BLOCKS_COL_PROXY:
+    /* proxy? */
+    tmp->proxy = atoi(tok);
+    break;
+
+  case BLOCKS_COL_SATTELLITEPROV:
+    /* sattelite provider? */
+    if (tok != NULL){
+    tmp->satprov = atoi(tok);
+    }
+    break;
+
+  case BLOCKS_COL_POSTAL:
+    /* postal code */
+    if (tok != NULL){
+      tmp->post_code = strndup(tok, strlen(tok));
+    }
+    else{
+      assert(tmp->post_code==NULL);
+    }
+    break;
+
+
+  case BLOCKS_COL_LAT:
+    /* latitude */
+    if (tok != NULL){
+    tmp->latitude = strtof(tok, &end);
+    if (end == tok || *end != '\0' || errno == ERANGE) {
+      ipmeta_log(__func__, "Invalid Latitude Value (%s)", tok);
+      state->parser.status = CSV_EUSER;
+      return;
+    }
+    }
+    break;
+
+  case BLOCKS_COL_LONG:
+    /* longitude */
+    if (tok != NULL){
+    tmp->longitude = strtof(tok, &end);
+    if (end == tok || *end != '\0' || errno == ERANGE) {
+      ipmeta_log(__func__, "Invalid Longitude Value (%s)", tok);
+      state->parser.status = CSV_EUSER;
+      return;
+    }
+    }
+    break;
+
+  case BLOCKS_COL_ACCURACY:
+    /* accuracy */
+    if (tok != NULL) {
+      tmp->accuracy = strtol(tok, &end, 10);
+      if (*tok != '\0' && (end == tok || *end != '\0' || errno == ERANGE)) {
+        ipmeta_log(__func__, "Invalid Area Code Value (%s)", tok);
+        state->parser.status = CSV_EUSER;
+        return;
+      }
+    }
+    break;
+
+  default:
+    ipmeta_log(__func__, "Invalid Blocks Column (%d:%d)", state->current_line,
+               state->current_column);
+    state->parser.status = CSV_EUSER;
+    break;
+  }
+
+  /* move on to the next column */
+  state->current_column++;
+}
+
+static void parse_blocks_row(int c, void *data)
+{
+  ipmeta_provider_t *provider = (ipmeta_provider_t *)data;
+  ipmeta_provider_maxmind_v2_state_t *state = STATE(provider);
+  /* Initializing block record*/
+  ipmeta_record_t *block_record = &(state->tmp_record);
+  /* the pointer to the final record */
+  ipmeta_record_t *record = NULL;
+  /*location record */
+  ipmeta_record_t *loc_record = NULL;
+  khiter_t khiter;
+
+
+  if (state->current_line < HEADER_ROW_CNT) {
+    state->current_line++;
+    return;
+  }
+
+  /* done processing the line */
+
+  /* make sure we parsed exactly as many columns as we anticipated */
+  if (state->current_column != BLOCKS_COL_COUNT) {
+    ipmeta_log(__func__,
+               "ERROR: Expecting %d columns in the blocks file, "
+               "but actually got %d",
+               BLOCKS_COL_COUNT, state->current_column);
+    state->parser.status = CSV_EUSER;
+    return;
+  }
+
+/* At the end of each row, look up the locations row in the hash table */
```

Comment:
it's probably worth writing a multi-line comment here explaining why we do this merge from location record into a (new) block record, and what the cost of this is